llama_index/indices/postprocessor/relevancy/base.py

```diff
@@ -0,0 +1,214 @@
+            "agg_mode": "intra_node", 
+        }, 
+        { 
+            "filter_type": "marginalization", 
+            "agg_mode": "intra_node",
+            "embedding_function": "default_embedding",
+            "similarity_function": "cosine",
+        }
+    ],
+    # node_synthesizers have access to the old nodes and the new nodes
+    "synthesizer": {
+        "score_mode": "new_avg",
+        "node_synthesizers": [
+            { 
+                "synthesizer_type": "summarizer", 
+                "op_mode": "sub_node",
+                "summarizer_type": "hugging-face",
+                "model_name": "t5-base", 
+            },
+            # Highlight the top 1 sentence, used in 
+            # relevancy highlighting formatter
+            { 
+                "synthesizer_type": "relevancy_highlights",
+                "op_mode": "intra_node", 
+                "relevancy_mode": "top_k", 
+                "top_k": 1
+            },
+            { 
+                "synthesizer_type": "formatter", 
+                "op_mode": "intra_node", 
+                "format_mode": "join",
+                "join_on": "...",
+            },
+        ]
+    },
+}
+"""
+
+
+class BaseRelevancyPostprocessor(BaseNodePostprocessor):
+    """
+    Base class for relevancy post-processors.
+
+    It is meant to encompass the stages of:
+    - Splitting / Chunking
+    - Ranking Function / Relevancy Scoring
+    - Filtering / Marginalization
+    - Node Synthesis: Aggregation, Mapping
+    """
+
+    filter_operator_shuffle: OperatorMode
+    scorer_operator_shuffle: OperatorMode = filter_operator_shuffle
+
+    _initial_nodes: Optional[List[NodeWithScores]] = None
+    _nodes: Optional[List[NodeWithScores]] = None
+    _subnodes: Optional[List[List[NodeWithScores]]] = None
+
+    filters = List[BaseFilter]
+    scorer: BaseRelevancyScorer  # default is to use old score...?
+
+    def __init__(self, config):
+        self.config = config
+
+    def validate(self):
+        if self.splitter is None:
+            assert self.scorer.agg_mode == AggMode.INTER_NODE
+            for filter in self.filters:
+                assert filter.agg_mode == AggMode.INTER_NODE
+            for synthesizer in self.node_synthesizers:
+                assert synthesizer.op_mode == MapMode.NODE
+        else:
+            assert self.scorer.agg_mode != AggMode.INTER_NODE
+            for filter in self.filters:
+                assert filter.agg_mode != AggMode.INTER_NODE
+            for synthesizer in self.node_synthesizers:
+                assert synthesizer.op_mode in [
+                    MapMode.SUB_NODE,
+                    MapMode.NODE,
+                    AggMode.INTRA_NODE,
+                ]
+
+    def postprocess_nodes(
+        self,
+        nodes: List[NodeWithScore],
+        query_bundle: Optional[QueryBundle] = None,
+    ) -> List[NodeWithScore]:
+        self._initial_nodes = nodes
+
+        if self.scorer.agg_mode == AggMode.INTER_NODE:
```

Comment:
This does suggest that there are essentially 2 major pathways, and within 1 of them, an additional 2 choices. I don't think that the `INTER/INTRA/CROSS_NODE` naming is particularly good UX, as it is an _imperative_ rather than name.

Reply:
Consider something like `COMPARE_NODES` / `COMPARE_SUBNODES`. Splitter only applies to the latter. We can still use this internal abstraction for the implementation, but rename the external interface to something more intuitive.

Reply:
Actually, the naming is just bad and confusing. Consider `SUBNODES_WITHIN_NODE`, `BETWEEN_NODES` and `SUBNODES_ACROSS_NODES`